src/cheri_pte.sail

```diff
@@ -13,14 +13,87 @@ bitfield PTE_Bits : pteAttribs = {
+ * capable of loading a tag may raise a trap even if the tag is not set).
+ * However, microarchitectures must gate their capability-load traps on the PTE
+ * bits (and sccsr.gclg[su]); that is, we permit relaxing only the dependence
+ * on the tag bit itself.
+ *
+ * While for StoreCap* the behaviors are as follows.
+ *
+ *   SC   SC_Mod  Action
+ *
+ *    0      0    Trap on tagged capability store
+ *    0      1    CAS the PTE to SC=1, SC_Mod=1 (i.e., set both bits)
+ *
+ *    1      0    [Reserved]
+ *    1      1    Permit tagged capability store
+ *
+ * It is permitted for implementations to ignore SC_Mod and behave as if it
+ * were always 0, at the expense of traps on cap-dirtying stores.  Note that SC
+ * and SC_Mod are not quite like W and D: SC is both the permission and dirty
+ * bit; SC_Mod just encapsulates a particular trap handler behavior.
+ * permitting SC_Mod to be treated as always zero means fewer things to
+ * initially implement.
+ *
+ * The SC=0 SC_Mod=1 behavior is described as a CAS of the PTE to close race
+ * conditions, much as with W and D.  Should the TLB have this state cached
+ * and then observe a SC=1 SC_Mod=1 PTE, no PTE write is necessary.  On the
+ * other hand, if SC=0 SC_Mod=0 is observed, the store operation must trap
+ * rather than transition the PTE to SC=1 SC_Mod=1 (or SC=1 SC_Mod=0).
+ *
+ * DATA DEPENDENCE: The data-dependent behavior of SC_Mod is difficult to model
+ * in the current sail RISC-V formal model in all cases: we here in the PTW are
+ * not given any more insight into the memory transaction than its type and, in
+ * particular, we cannot see the stored tag of a Write(Cap) or ReadWrite(Cap)
+ * transaction. Sore instructions can somewhat abusively use the tag to
+ * discriminate between Write(Cap) and Write(Data) transaction types even when
+ * the instruction itself is capable of storing a capability.  However, for
+ * ReadWrite(Cap), there is no easy way to say what we might mean.  Therefore,
+ * we take any Cap-storing transaction as grounds to mark the page dirty, even
+ * if no capability was transferred. (In fact, in the case of ReadWrite(Cap),
+ * the load may fault due to LC_Mod gating, and so perhaps no store transaction
+ * takes place at all; even in this case, SC_Mod may still assert SC!) Other
+ * microarchitectural implementations are free to behave similarly or to
+ * provide the narrower interpretation of SC_Mod and require that the tag bit
+ * actually be set before dirtying the page.
+ *
+ * SV32: There are no extension bits available, so we hard-code the result to
+ * be LC=1 LC_Mod=0 LC_Gen=0 (loads succeed and preserve tags, LCG mechanism
+ * not enabled) and SC=1 SC_Mod=1 (stores succeed).
  */
 type extPte = bits(10)
 
 bitfield Ext_PTE_Bits : extPte = {
-  LoadCap  : 9,
-  StoreCap : 8,
+  StoreCap     : 9, /* Permit capability stores */
+  LoadCap      : 8, /* Permit capability loads */
+  StoreCap_Mod : 7, /* Modify capability store inhibit; see above table */
```

Comment:
The use of inhibit here and in the line below is perhaps stale now?

Reply:
Yup, good eye.